File: frontend/src/types/deal.ts
// frontend/src/types/deal.ts - Tipos TypeScript para Deals (Phase 3)

// Enums para Deal (MAYÚSCULAS para coincidir con backend)
export type DealStage = 'LEAD' | 'CONTACTED' | 'QUALIFIED' | 'PROPOSAL' | 'NEGOTIATION' | 'WON' | 'LOST'

export type DealPriority = 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT'

export type DealSource =
  | 'WEBSITE'
  | 'REFERRAL'
  | 'COLD_CALL'
  | 'SOCIAL_MEDIA'
  | 'EMAIL_CAMPAIGN'
  | 'EVENT'
  | 'PARTNER'
  | 'OTHER'

// Interface base para Deal
export interface DealBase {
  title: string
  description: string | null
  value: number
  currency: string
  stage: DealStage
  probability: number
  priority: DealPriority
  source: DealSource | null
  expected_close_date: string | null
  actual_close_date: string | null
  lost_reason: string | null
  tags: string[]
  custom_fields: Record<string, any>
  is_active: boolean
}

// Interface para crear Deal
export interface DealCreate extends DealBase {
  contact_id: string
}

// Interface para actualizar Deal
export interface DealUpdate extends Partial<DealBase> {}

// Interface para Deal completo (respuesta del servidor)
export interface DealOut extends DealBase {
  id: string
  contact_id: string
  owner_id: number
  created_at: string
  updated_at: string
  // Opcional: información del contacto y owner cuando se incluyen
  contact?: {
    id: string
    full_name: string
    email?: string | null
    phone?: string | null
    mobile?: string | null
    company?: string | null
  }
  owner?: {
    id: number
    full_name: string
    email: string
  }
}

// Interface para Deal con información del contacto
export interface DealWithContact extends DealOut {
  contact_name: string | null
  contact_email: string | null
  contact_company: string | null
}

// Interface para respuesta paginada
export interface DealsPaginatedResponse {
  items: DealOut[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

// Interface para estadísticas de deals
export interface DealStats {
  total_deals: number
  active_deals: number
  won_deals: number
  lost_deals: number
  total_value: number
  won_value: number
  potential_value: number
  average_deal_value: number
  average_win_probability: number
  deals_by_stage: Record<DealStage, number>
  value_by_stage: Record<DealStage, number>
  deals_by_priority: Record<DealPriority, number>
  deals_by_source: Record<DealSource, number>
  win_rate: number
  conversion_rate: number
}

// Interface para actualizar etapa
export interface DealStageUpdate {
  stage: DealStage
  probability?: number
  lost_reason?: string
}

// Interface para resumen del pipeline
export interface PipelineStageData {
  stage: DealStage
  count: number
  total_value: number
  deals: DealOut[]
}

export type PipelineSummary = Record<DealStage, PipelineStageData>

// Labels para las etapas
export const StageLabels: Record<DealStage, string> = {
  LEAD: 'Lead',
  CONTACTED: 'Contactado',
  QUALIFIED: 'Calificado',
  PROPOSAL: 'Propuesta',
  NEGOTIATION: 'Negociación',
  WON: 'Ganado',
  LOST: 'Perdido',
}

// Labels para prioridades
export const PriorityLabels: Record<DealPriority, string> = {
  LOW: 'Baja',
  MEDIUM: 'Media',
  HIGH: 'Alta',
  URGENT: 'Urgente',
}

// Labels para fuentes
export const SourceLabels: Record<DealSource, string> = {
  WEBSITE: 'Sitio Web',
  REFERRAL: 'Referido',
  COLD_CALL: 'Llamada en Frío',
  SOCIAL_MEDIA: 'Redes Sociales',
  EMAIL_CAMPAIGN: 'Campaña Email',
  EVENT: 'Evento',
  PARTNER: 'Socio/Partner',
  OTHER: 'Otro',
}

// Opciones para selects
export const STAGE_OPTIONS = Object.entries(StageLabels).map(([value, label]) => ({
  value: value as DealStage,
  label,
}))

export const PRIORITY_OPTIONS = Object.entries(PriorityLabels).map(([value, label]) => ({
  value: value as DealPriority,
  label,
}))

export const SOURCE_OPTIONS = Object.entries(SourceLabels).map(([value, label]) => ({
  value: value as DealSource,
  label,
}))

// Colores para etapas (para UI)
export const StageColors: Record<DealStage, string> = {
  LEAD: 'gray',
  CONTACTED: 'blue',
  QUALIFIED: 'cyan',
  PROPOSAL: 'purple',
  NEGOTIATION: 'orange',
  WON: 'green',
  LOST: 'red',
}

// Colores para prioridades
export const PriorityColors: Record<DealPriority, string> = {
  LOW: 'gray',
  MEDIUM: 'blue',
  HIGH: 'orange',
  URGENT: 'red',
}

// Probabilidades sugeridas por etapa
export const StageProbabilities: Record<DealStage, number> = {
  LEAD: 10,
  CONTACTED: 20,
  QUALIFIED: 40,
  PROPOSAL: 60,
  NEGOTIATION: 80,
  WON: 100,
  LOST: 0,
}

File: backend/app/schemas/deal.py
# backend/app/schemas/deal.py - Schemas para Deals (Phase 3)

from pydantic import BaseModel, Field, field_validator
from datetime import datetime, date
from typing import Optional, List

from app.models.deal import DealStage, DealPriority, DealSource


# Schema base para Deal
class DealBase(BaseModel):
    title: str = Field(..., min_length=2, max_length=300, description="Título del deal")
    description: Optional[str] = Field(None, description="Descripción detallada")
    value: float = Field(default=0.0, ge=0, description="Valor monetario del deal")
    currency: str = Field(default="USD", max_length=10, description="Moneda (USD, EUR, etc)")
    stage: DealStage = Field(default=DealStage.LEAD, description="Etapa del pipeline")
    probability: int = Field(default=0, ge=0, le=100, description="Probabilidad de cierre (%)")
    priority: DealPriority = Field(default=DealPriority.MEDIUM, description="Prioridad")
    source: Optional[DealSource] = Field(None, description="Fuente de origen")
    expected_close_date: Optional[date] = Field(None, description="Fecha estimada de cierre")
    actual_close_date: Optional[date] = Field(None, description="Fecha real de cierre")
    lost_reason: Optional[str] = Field(None, max_length=500, description="Motivo de pérdida")
    tags: List[str] = Field(default_factory=list, description="Tags del deal")
    custom_fields: dict = Field(default_factory=dict, description="Campos personalizados")
    is_active: bool = Field(default=True, description="Estado activo")

    @field_validator("stage", "priority", "source", mode="before")
    @classmethod
    def normalize_enums(cls, v):
        """Normaliza enums a mayúsculas"""
        if isinstance(v, str):
            return v.upper()
        return v


# Schema para crear Deal
class DealCreate(DealBase):
    contact_id: int = Field(..., description="ID del contacto asociado")

    @field_validator('probability')
    @classmethod
    def validate_probability(cls, v):
        if v < 0 or v > 100:
            raise ValueError('La probabilidad debe estar entre 0 y 100')
        return v

    @field_validator('value')
    @classmethod
    def validate_value(cls, v):
        if v < 0:
            raise ValueError('El valor no puede ser negativo')
        return v


# Schema para actualizar Deal
class DealUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=2, max_length=300)
    description: Optional[str] = None
    value: Optional[float] = Field(None, ge=0)
    currency: Optional[str] = Field(None, max_length=10)
    stage: Optional[DealStage] = None
    probability: Optional[int] = Field(None, ge=0, le=100)
    priority: Optional[DealPriority] = None
    source: Optional[DealSource] = None
    expected_close_date: Optional[date] = None
    actual_close_date: Optional[date] = None
    lost_reason: Optional[str] = Field(None, max_length=500)
    tags: Optional[List[str]] = None
    custom_fields: Optional[dict] = None
    is_active: Optional[bool] = None

    @field_validator("stage", "priority", "source", mode="before")
    @classmethod
    def normalize_enums(cls, v):
        """Normaliza enums a mayúsculas"""
        if isinstance(v, str):
            return v.upper()
        return v


# Schema para respuesta de Deal
class DealOut(DealBase):
    id: int
    contact_id: int
    owner_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Schema para Deal con información del contacto
class DealWithContact(DealOut):
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None
    contact_company: Optional[str] = None


# Schema para paginación de Deals
class DealsPaginatedResponse(BaseModel):
    items: List[DealOut]
    total: int
    page: int
    page_size: int
    total_pages: int


# Schema para estadísticas de Deals
class DealStats(BaseModel):
    # Totales
    total_deals: int
    active_deals: int
    won_deals: int
    lost_deals: int

    # Valor total
    total_value: float
    won_value: float
    potential_value: float

    # Promedio
    average_deal_value: float
    average_win_probability: float

    # Por etapa
    deals_by_stage: dict  # {stage: count}
    value_by_stage: dict  # {stage: total_value}

    # Por prioridad
    deals_by_priority: dict  # {priority: count}

    # Por fuente
    deals_by_source: dict  # {source: count}

    # Conversión
    win_rate: float  # Porcentaje de deals ganados
    conversion_rate: float  # Porcentaje de leads convertidos


# Schema para mover deal en el pipeline
class DealStageUpdate(BaseModel):
    stage: DealStage
    probability: Optional[int] = Field(None, ge=0, le=100)
    lost_reason: Optional[str] = Field(None, max_length=500)
